data/rule/rule.go

```diff
@@ -0,0 +1,117 @@
+// Package rule contains rule-based heuristic implementations.
+// It is used in the generated code in content.go for disambiguation of languages
+// with colliding extensions based on regexps from Linguist data.
+package rule
+
+import (
+	"regexp"
+)
+
+// Heuristic consist of a number of rules where each, if matches,
+// identifes content as belonging to a programming language(s).
+type Heuristic interface {
+	Matcher
+	Languages() []string
+}
+
+// Matcher checks if the data matches (number of) pattern.
+// Every rule below implements this interface: a rule is matcher that identifies
+// given programming language(s) in case of the match.
+type Matcher interface {
+	Match(data []byte) bool
+}
+
+// languages struct incapsulate data common to every Matcher: all languages
+// that it identifies.
+type languages struct {
```

Comment:
Since this type is returned by the exported members of this package, we should probably export it as well. Otherwise it can be annoyingly inconvenient to use.

Reply:
(It's not a hard-and-fast rule, but generally a good idea for exported functions not to return unexported types)

Reply:
Left private as this is designed to be embedded and the only function that returns it is a factory method that is used for constructing a data-structure literal.